File: 240522/최소 경로로 탈출 하기/escape-with-min-distance.js
class Queue {
    
    constructor() {
        this.q = []
        this.front = -1
        this.tail = -1
    }

    push(item) {
        this.q.push(item)
        this.tail++
    }

    empty() {
        return this.front === this.tail
    }

    pop() {
        if (this.empty()) {
            throw new Error("empty")
        }
        return this.q[++this.front]
    }

    size() {
        return this.tail - this.front
    }
}

const fs = require('fs')
const input = fs.readFileSync(0).toString().trim().split('\n')

const [n, m] = input[0].split(" ").map(Number)
const arr = input.slice(1, n+1).map(x => x.split(" ").map(Number))
const visited = Array.from(new Array(n), () => Array(m).fill(0))

dir = [[0, 1], [1, 0], [0, -1], [-1, 0]]
function bfs() {
    q = new Queue
    q.push([0, 0])
    visited[0][0] = 1

    while (q.size() > 0) {
        const [x, y] = q.pop()

        dir.forEach(d => {
            xd = d[0] + x
            yd = d[1] + y

            if (0 <= xd && xd < n && 0 <= yd && yd < m) {
                if (arr[xd][yd] == 1 && visited[xd][yd] == 0) {
                    visited[xd][yd] = visited[x][y] + 1
                    q.push([xd, yd])
                }
            }
        })
    }
}

bfs()

ans = visited[n-1][m-1]
if (ans) {
    console.log(ans-1)
} else {
    console.log(-1)
}

File: 240522/최소 경로로 탈출 하기/escape-with-min-distance.py
n, m = map(int, input().split())
arr = [list(map(int, input().split())) for _ in range(n)]
visited = [[0 for _ in range(m)] for _ in range(n)]

def bfs():
    q = [(0, 0)]
    dir = [(0, 1), (1, 0), (-1, 0), (0, -1)]
    visited[0][0] = 1
    while q:
        x, y = q.pop(0)
     
        for dn, dm in dir:
            xdn, ydm = x + dn, y + dm
            if 0 <= xdn < n and 0 <= ydm < m:
                if visited[xdn][ydm] == 0 and arr[xdn][ydm] == 1:
                    visited[xdn][ydm] += visited[x][y] + 1
                    q.append((xdn, ydm))


bfs()
if visited[n-1][m-1]:
    print(visited[n-1][m-1]-1)
else:
    print(-1)